### PR TITLE
Replace usage of nccmp with internal netCDF comparison tool

### PR DESCRIPTION
**Pull request by tjtg**
The acceptance tests currently use `nccmp`, which is an external dependency written in C and requiring separate installation outside of the Python environment used for the rest of IMPROVER.
This PR adds an internal tool for comparing netCDF files and replaces usage of `nccmp` in the acceptance tests with this internal tool.

Also note that nccmp expresses relative tolerances as percentages, which is different to numpy.allclose combined relative and absolute tolerance method, and expression of relative tolerances as floats.
Adjusting the tolerance constants to be similar to previous causes two estimate_emos_coefficients tests to fail and inspection of the output shows differences which may be scientifically significant. I have marked these two tests as xfail in this branch.
These different EMOS coefficient outputs will need to be investigated and scientifically appropriate tolerances determined for these tests.

Testing:
 - [x] Ran tests and they passed OK
 - [x] Added new tests for the new feature(s)

### CODE REVIEW

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- lib/improver/utilities/compare.py  10
- lib/improver/cli/compare.py  1
- lib/improver/tests/utilities/test_compare.py  9
- lib/improver/tests/acceptance/test_compare.py  4
         

Complexity decreasing per file
==============================
+ lib/improver/tests/acceptance/acceptance.py  -2
         

Clones added
============
- lib/improver/tests/utilities/test_compare.py  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=4768470&bid=15893567)

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- lib/improver/utilities/compare.py  10
- lib/improver/tests/utilities/test_compare.py  9
- lib/improver/tests/acceptance/test_compare.py  4
- lib/improver/cli/compare.py  1
         

Complexity decreasing per file
==============================
+ lib/improver/tests/acceptance/acceptance.py  -2
         

Clones added
============
- lib/improver/tests/utilities/test_compare.py  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=4768470&bid=15893567)

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- lib/improver/utilities/compare.py  10
- lib/improver/tests/acceptance/test_compare.py  4
- lib/improver/tests/utilities/test_compare.py  9
- lib/improver/cli/compare.py  1
         

Complexity decreasing per file
==============================
+ lib/improver/tests/acceptance/acceptance.py  -2
         

Clones added
============
- lib/improver/tests/utilities/test_compare.py  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=4768470&bid=15893567)

File: lib/improver/tests/acceptance/acceptance.py [hunk 1]
```diff
@@ -228,12 +139,8 @@ def compare(output_path, kgo_path, recreate=True, **kwargs):
...
         kgo_path (pathlib.Path): Path to KGO file
         recreate (bool): False to disable KGO recreation, compare only
-
-    Keyword Args:
-        exclude_dims (Optional[List[str]]): dimensions to exclude
-        options (Optional[str]): comparison options
-        atol (Optional[float]): absolute tolerance
-        rtol (Optional[float]): relative tolerance
+        atol (float): absolute tolerance
+        rtol (float): relative tolerance
```
Comment (bayliffe) [inline lib/improver/tests/acceptance/acceptance.py]: ```exclude_vars``` needs to be defined here.

File: lib/improver/tests/acceptance/test_compare.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+def test_different(capsys):
+    """Compare different files, should report differences to stdout"""
+    kgo_dir = acc.kgo_root()
+    a_file = kgo_dir / "percentile/basic/input.nc"
+    b_file = kgo_dir / "threshold/basic/input.nc"
+    args = [a_file, b_file]
+    run_cli(args)
+    captured = capsys.readouterr()
+    print(captured.out)
```
Comment (bayliffe) [inline lib/improver/tests/acceptance/test_compare.py]: extraneous print statement.

File: lib/improver/tests/acceptance/acceptance.py (hunk 1 above)
Comment (tjtg) [reply lib/improver/tests/acceptance/acceptance.py]: I've added this to the docstring.

File: lib/improver/tests/acceptance/test_compare.py (hunk 2 above)
Comment (tjtg) [reply lib/improver/tests/acceptance/test_compare.py]: I've removed this `print`, as well as a few others in the newly added tests.

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- lib/improver/tests/acceptance/test_compare.py  4
- lib/improver/cli/compare.py  1
- lib/improver/utilities/compare.py  10
- lib/improver/tests/utilities/test_compare.py  9
         

Complexity decreasing per file
==============================
+ lib/improver/tests/acceptance/acceptance.py  -2
         

Clones added
============
- lib/improver/tests/utilities/test_compare.py  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=4768470&bid=15893567)

File: lib/improver/utilities/compare.py [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+            if isinstance(desired_attr, np.ndarray):
+                if not np.array_equal(actual_attr, desired_attr):
+                    msg = (f"different attribute value {name}/{key} - "
+                           f"{actual_attr} {desired_attr}")
+                    reporter(msg)
+            elif actual_attr != desired_attr:
+                msg = (f"different attribute value {name}/{key} - "
+                       f"{actual_attr} {desired_attr}")
+                reporter(msg)
+        except KeyError:
```
Comment (benfitzpatrick) [inline lib/improver/utilities/compare.py]: Do we expect a KeyError here?

File: lib/improver/cli/compare.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+    Returns:
+        None
+    """
+
+    def print_reporter(message):
+        print(message)
+
+    compare.compare_netcdfs(actual, desired, rtol=rtol, atol=atol,
+                            reporter=print_reporter)
```
[the remark this answers is not in the document]
Comment (tjtg) [reply lib/improver/cli/compare.py]: There was originally some other code in this function to keep a count of the number of differences identified and exit with zero/non-zero (similar to `grep` or `diff` tools), but the return value from the function wasn't being passed through to an exit status number.
I've changed this to `reporter=print` now.

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- lib/improver/tests/acceptance/test_compare.py  4
- lib/improver/cli/compare.py  1
- lib/improver/tests/utilities/test_compare.py  9
- lib/improver/utilities/compare.py  10
         

Complexity decreasing per file
==============================
+ lib/improver/tests/acceptance/acceptance.py  -2
         

Clones added
============
- lib/improver/tests/utilities/test_compare.py  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=4768470&bid=15893567)

File: lib/improver/utilities/compare.py (hunk 3 above)
Comment (tjtg) [reply lib/improver/utilities/compare.py]: A KeyError will occur when `actual_ds` is missing an attribute key which exists in `desired_ds` and line 240 attempts to access it. The different attribute keys will be reported earlier on line 236, but the comparison will continue in order to try and produce a more complete list of differences (assuming the reporter function doesn't raise/exit and end the comparison early)

**Pull request merged**